syntaxes/src/expression.ts

```diff
@@ -0,0 +1,892 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import { GrammarDefinition } from './types';
+
+export const Expression: GrammarDefinition = {
+  scopeName: 'expression.ng',
+  injectionSelector: 'L:text.html -comment',
+  patterns: [{include: '#ngExpression'}],
+  repository: {
+    ngExpression: {
+      name: 'meta.expression.ng',
+      patterns: [
+        {
+          include: '#string',
+        },
+        {
+          include: '#literal',
+        },
+        {
+          include: '#ternaryExpression',
+        },
+        {
+          include: '#expressionOperator',
+        },
+        {
+          include: '#functionCall',
+        },
+        {
+          include: '#identifiers',
+        },
+        {
+          include: '#parenExpression',
+        },
+        {
+          include: '#punctuationComma',
+        },
+        {
+          include: '#punctuationAccessor',
+        },
+      ],
+    },
+
+    arrayLiteral: {
+      name: 'meta.array.literal.ts',
+      begin: /\[/,
+      beginCaptures: {
+        0: {
+          name: 'meta.brace.square.ts',
+        },
+      },
+      end: /\]/,
+      endCaptures: {
+        0: {
+          name: 'meta.brace.square.ts',
+        },
+      },
+      patterns: [
+        {
+          include: '#ngExpression',
+        },
+        {
+          include: '#punctuationComma',
+        },
+      ],
+    },
+
+    booleanLiteral: {
+      patterns: [
+        {
+          name: 'constant.language.boolean.true.ts',
+          match: /(?<!\.|\$)\btrue\b(?!\$)/,
+        },
+        {
+          name: 'constant.language.boolean.false.ts',
+          match: /(?<!\.|\$)\bfalse\b(?!\$)/,
+        },
+      ],
+    },
+
+    expressionOperator: {
+      patterns: [
+        {
+          match: /((?<!\|)(\|)(?!\|))\s?([a-zA-Z0-9\-\_\$]*)/,
+          captures: {
+            2: {
+              name: 'keyword.operator.logical.ts',
+            },
+            3: {
+              name: 'entity.name.function.pipe.ng',
+            },
+          },
+        },
```

Comment:
If you remove the parens around the `\|` in the middle of the first group, it won't create a duplicate group and then you can change the `captures` keys to `1` and `2` instead of `2` and `3`. Compare the Details tab [here](https://regexr.com/4tk7o) and [here](https://regexr.com/4tk7r) with the match selected.

Reply:
Removing parenthesis from the negative lookeahead and negative lookbehind invalidate it. (At least in [Rubular](https://rubular.com/) - I use Rubular to test my regexes because it is made using Ruby, and Oniuruma is the Ruby regex engine).

But removing the enclosing parenthesis from the lookbehind, `|` capture group and lookahead makes it capture only once.

`((?<!\|)(\|)(?!\|))\s?([a-zA-Z0-9\-\_\$]*)` became  `(?<!\|)(\|)(?!\|)\s?([a-zA-Z0-9\-\_\$]*)`

Reply:
Good call! Done that